### ISSUE TITLE
Allow modifications of sequence in SFF objects

### ISSUE THREAD

**Issue by twrightsman** [biopython/biopython#880]
Migrated from https://redmine.open-bio.org/issues/3438 (*Note - don't write # number as GitHub will auto-link that as if it was a GitHub issue; Peter*)

Original discussion from @mmokrejs:

> I find it a bit awkward but namely, with unnecessary overhead, to edit a sequence in a SFF object. I am fine with the requirement that the length of the sequence must stay same is the length of qualities and other annotation lists. However, to edit a sequence I have to do now:

```
    if _was_modified:
        _letter_annotations = _record.letter_annotations
        _annotations = _record.annotations
        _record.letter_annotations = {}
        _record.annotations = {}
        _record.seq = Seq(_sequence, generic_dna)
        _record.letter_annotations = _letter_annotations
        _record.annotations = _annotations

        _new_record = SeqRecord(Seq(_sequence, generic_dna), id=_record.id, name=_record.name, description=_record.description, annotations=_record.annotations, letter_annotations=_record.letter_annotations)

        _wrote = SeqIO.write(_new_record, _fh, 'sff')
    else:
        _wrote = SeqIO.write(_record, _fh, 'sff')
```

> The whole work in backup&restore of annotation lists is not necessary in my eyes. I think providing _record.rewrite_sequence('tcagnnnnnnnn') would be quite helpful here.

**Comment by twrightsman** [comment]
Doesn't seem to be resolved:

```
$ python
Python 2.7.12 (default, Jun 29 2016, 14:05:02) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from Bio import SeqIO
>>> from Bio.Seq import Seq
>>> record = SeqIO.read('SRR088776_short_F95S1KP01DA6OU.sff', 'sff')
/usr/local/lib/python2.7/site-packages/Bio/SeqIO/SffIO.py:671: BiopythonParserWarning: Overlapping clip values in SFF record
  warnings.warn("Overlapping clip values in SFF record", BiopythonParserWarning)
/usr/local/lib/python2.7/site-packages/Bio/SeqIO/SffIO.py:988: BiopythonParserWarning: Your SFF file is technically invalid as it is missing a terminal 4 byte null padding region.
  BiopythonParserWarning)
>>> dir(record)
['__add__', '__bool__', '__class__', '__contains__', '__delattr__', '__dict__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__iter__', '__le___', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__nonzero__', '__radd__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_per_letter_annotations', '_seq', '_set_per_letter_annotations', '_set_seq', 'annotations', 'dbxrefs', 'description', 'features', 'format', 'id', 'letter_annotations', 'lower', 'name', 'reverse_complement', 'seq', 'upper']
>>> new_seq = str(record.seq).replace('n', 'c')
>>> record.seq = Seq(new_seq)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/Bio/SeqRecord.py", line 301, in _set_seq
    raise ValueError("You must empty the letter annotations first!")
ValueError: You must empty the letter annotations first!
>>> 
```

**Comment by peterjc** [comment]
Yes, this is a usability design issue - we enforce the per-letter-annotations dictionary entries being the same length as the sequence, which makes changing them both together tricky.

Currently you are expected to do clear this before you can change the sequence:

```
record.letter_annotations = {}
record.seq = Seq(new_seq)
```

I'm not convinced that adding a new method would actually help here. One alternative would be to allow the new string assignment, and if the length differs, drop the per-letter-annotations with a warning.

**Comment by mmokrejs** [comment]
Well, as I show in my code I of course do not want to discard any of the annotations, and in the code I copy them all back. The whole workaround just obeys biopython's roadblocks so I could edit the sequence, and in my eyes it would be nice to do just:

_record.rewrite_sequence('tcagnnnnnnnn')

Fine with me, the code should require length of the sequence to be same as of the list in annotation dictionaries. I can live without this, but provided one cannot slice or edit SFF objects at all, this would be a nice although small step forward.

> and if the length differs, drop the per-letter-annotations with a warning.
Better not, maybe an optional argument to override the blocker, but no discards of any data please. I know we could easily run into a problem with too many Warnings, so moving to WarnOnce, etc. I think it is better to avoid this and eventually provide a swithc to override the default deny if lengths differ. But this is certainly a different thing than I asked for, so do not worry about this. ;)

**Comment by peterjc** [comment]
Here's a concrete example:

```
>>> from Bio.Alphabet import generic_dna
>>> from Bio.Seq import Seq
>>> from Bio import SeqIO
>>> record_dict = SeqIO.index("Tests/Roche/E3MFGYR02_random_10_reads.sff", "sff")
>>> record = record_dict["E3MFGYR02GFKUC"]
>>> print(len(record))
299
```

If replacing the sequence, this typically would invalidate the old per-letter-annotations. So, to avoid accidental data loss by design I force you to do this explicitly:

```
>>> record.letter_annotations = {}
>>> record.seq = Seq("ACGT", generic_dna)
>>> record.letter_annotations["phred_quality"] = [40, 40, 40, 40]
```

There is a special case where perhaps the ``ValueError`` is not needed, which is when the new sequence is the same length (e.g. you are editing the case of the letters, or masking some of them), e.g.

```
>>> record.seq = Seq("N" * len(record), generic_dna)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "Bio/SeqRecord.py", line 301, in _set_seq
    raise ValueError("You must empty the letter annotations first!")
ValueError: You must empty the letter annotations first!
```

@mmokrejs Are you replacing the sequence with one of the same length or not?

**Comment by mmokrejs** [comment]
> Are you replacing the sequence with one of the same length or not?

Sorry Peter for being unclear here, yes, I just modify the sequence while the length is unchanged. The code I posted initially does not change length of the sequence either. Therefore

`record.letter_annotations = {}`

can be avoided and the ValueError is really more a nuisance than a necessary check.

**Comment by mmokrejs** [comment]
Thank you Peter, I just lowercased the whole sequence in a SFF file or replaced with all n's. I also exported the SFF objects into SFF and FASTQ, and compared the SFF contents between the old and new SFF file: only the nucleotide sequence changed, good. I conclude it works as I wanted. ;)

Than I event tested whether I can manipulate PHREDQ values, and indeed I can:
```
_record.seq = len(_record.seq) * "n"
_record.letter_annotations['phred_quality']=len(_record.letter_annotations['phred_quality']) * [5]
SeqIO.write(_record, _fh, "sff")
SeqIO.write(_record, _fh2, "fastq")
```

Tested on top of biopython-1.67.
